Contributing Guide (#15)
infobloxopen/universal-ddi-ansible: docs/CONTRIBUTING.md

# Contributing to infoblox.universal_ddi collection

Welcome to the Universal DDI Ansible Repository! We are excited that you are interested in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure.

## Workflow Summary

1. [Clone the repository](#cloning)
2. [Make the desired Code Change](#writing-new-code)
3. [Run integration tests locally and ensure that they pass](#writing-integration-tests)
4. [Create a PR](#pr-creation)

## Cloning

The `ansible-test` command expects that the repository is in a directory that matches it's collection,
under a directory `ansible_collections`. Clone ensuring that hierarchy:

```shell
mkdir -p $TARGET_DIR/ansible_collections/infoblox
git clone <url> $TARGET_DIR/ansible_collections/infoblox/universal_ddi
```

The python requirements must be installed separately. It is recommended to use a venv to keep the dependencies local. 
Set up your Python virtual environment

```shell
cd $TARGET_DIR/ansible_collections/infoblox/universal_ddi
python3 -m venv .venv
source .venv/bin/activate
pip3 install ansible

> pip3 install -r requirements.txt

Reviewer: Need to add one more command 
` pip install -r test-requirements.txt `

```

You can run ```ansible-test sanity``` to test if your collection is ready for development

> ## Writing New Code

Reviewer: ```suggestion
## Implementing the code changes
```

> ## Writing New Code

Reviewer: Can you add a section or a sub section before this , giving the link to the Bloxone API documentation stating that the user can refer to this link for all information and attributes related to the module they want to contribute ?

> ### Writing New Modules

Reviewer: ```suggestion
### Developing New Modules
```

 
Create two files for each module:

> - `module_name.py`: The main module code.

Reviewer: ```suggestion
- `module_name.py`: Module for performing CRUD operations
```

> - `module_name_info.py`: For gathering module resource information.

Reviewer: ```suggestion
- `module_name_info.py`: Module to List / Read the object
```

Place these files under `/plugins/modules`.

#### Documentation Examples and Return section inside modules

**Documentation Section** Each module file should start with a comprehensive documentation section. [Module format and documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#module-format-and-documentation) guide gives  details about how to document.

**Examples Section** Provides practical usage examples demonstrating standard and edge case scenarios.

> **Return Section** Clearly document what values the module returns.

Reviewer: ```suggestion
**Return Section** -  It defines what data is supposed to be returned and how it is structured.
```

> ### Writing Integration Tests

Reviewer: Can you shift this section to testing the code ? 
This way we can separate development and testing.

- All modules MUST have integration tests for new features. Bug fixes for modules that currently have integration tests SHOULD have tests added.

- Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature.

- Each module must also be added to the all group in meta/runtime.yml. This will allow to reuse common variables like portal_key in the test. 

#### Expected test criteria:
- Resource creation under check mode
- Resource creation
- Resource creation again (idempotency)
- Resource deletion under check mode
- Resource deletion
- Resource deletion again (idempotency)

> - Resource creation with additional parameters

Reviewer: ```suggestion
- Resource updation with all additional attributes under the object
```

> - Resource modification

Reviewer: Not Required as such , the line above would cover this

> 

Reviewer: https://github.com/IBM/z_ansible_collections_samples/blob/main/docs/share/contribution-guidelines.md#linting

> #### Running Tests Locally

Reviewer: ```suggestion
#### Testing the code changes
```

> For local testing, ensure you set up your integration test configuration

Reviewer: Can you place this under a sub section of `Testing the Code Changes ` , naming it either `Configuration` .

1. Navigate to `tests/integration` folder
2. Create the integration_config.yaml file with necessary environment-specific configurations such as portal_url and portal_key as shown below.

>    ```yaml

Reviewer: Add the module_defaults part we have in the integration tests and also add a line stating that we define the configuration details under universal_ddi/all so the config values get added to all the calls in the playbook 
(please rephrase it )

   portal_url: https://csp.infoblox.com
   portal_key: "your-portal-secret-key"
   ```
   - The default URL for the Cloud Services Portal is https://csp.infoblox.com
   - An API key is required to access Infoblox API. You can obtain an API key by following the instructions in the guide for [Configuring User API Keys](https://docs.infoblox.com/space/BloxOneCloud/35430405/Configuring+User+API+Keys).

Run tests for the desired module(s):

```shell
ansible-test integration <module_name> 
```
If you want more detailed output, run the command with -vvv 

```shell
ansible-test integration <module_name> -vvv
```

## PR Creation

- Before raising your PR, you must make sure the PR is merge worthy.
- The PR will trigger some check actions in Github. The PR can be only merged when these tests pass. You can also test your PR locally with these commands. 

```shell
ansible-test sanity
```

```shell
ansible-test integration
```

> 

Reviewer: One more line about how the PR would go under review by maintainers of this repository and how the contributor is expected to make changes as per the review for the PR to get merged.

### Developing New Modules
 
Create two files for each module:
- `module_name.py`: Module for performing CRUD operations.
- `module_name_info.py`: Module to List / Read the object.

Place these files under `/plugins/modules`.

#### Documentation Examples and Return section inside modules

**Documentation Section** Each module file should start with a comprehensive documentation section. [Module format and documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#module-format-and-documentation) guide gives  details about how to document.

**Examples Section** Provides practical usage examples demonstrating standard and edge case scenarios.

**Return Section** It defines what data is supposed to be returned and how it is structured.

## Testing the Code Changes

### Writing Integration Tests

- All modules MUST have integration tests for new features. Bug fixes for modules that currently have integration tests SHOULD have tests added.

> - Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature.

Reviewer: ```suggestion
- Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature. The tests are supposed to be added for all writable attributes of the object. The complete list of attributes can be found in the API documentation.
```

> - Each module must also be added to the all group in meta/runtime.yml. This will allow to reuse common variables like portal_key in the test. 

Reviewer: Make this the first point of this section

### Expected test criteria

- Resource creation under check mode
- Resource creation

> - Resource creation again (idempotency)

Reviewer: Removing the delete wasnt necessary , we still need to mention that 
Delete Check mode , delete and delete idempotent are required.

Also adda a line that for a more detailed example , the contributor can one at one of the examples in the tests section.

Reviewer: To make it better , this can be referred 
https://github.com/openstack/ansible-collections-openstack/blob/master/docs/contributing.md#examples

# Contributing to infoblox.universal_ddi collection

Welcome to the Universal DDI Ansible Repository ! We're thrilled about your interest in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure and coding standards.
## Workflow Summary

1. [Setup](#setup)
2. [Implementing the Code Changes](#implementing-the-code-changes)
3. [Testing the Code Changes](#testing-the-code-changes)
4. [Creating a PR](#creating-a-pr)

## Setup

> 

Reviewer: Lets divide this setup in two segments 
Directory Structure and Installing requirements

The `ansible-test` command expects that the repository is in a directory that matches it's collection. Here's the required directory structure you need to set up in your development environment:

```
<target_directory>/ansible_collections/infoblox/universal_ddi/
├── changelogs/
├── docs/
├── meta/
├── plugins/
│   ├── doc_fragments/
│   ├── lookup/
│   ├── module_utils/
│   ├── modules/
│   └── plugin_utils/
├── tests
│   ├── integration/
│   │   └── targets/
│   └── unit/
└── venv/
```
Before making any changes, it's essential to fork the repository. This allows you to work on your own copy of the project and makes it easy to contribute your changes back to the main repository through a pull request.

Clone the repository from GitHub ensuring the hierarchy.
```shell
mkdir -p $TARGET_DIR/ansible_collections/infoblox
git clone https://github.com/<your_username>/universal-ddi-ansible.git $TARGET_DIR/ansible_collections/infoblox/universal_ddi
```

> 

Reviewer: The Installing requirements can start from here . 
Also we should separate the setting up of venv and installation code into different code blocks , as some users might already have their venv setup .

   ```yaml
      - module_defaults:
          group/infoblox.universal_ddi.all:
            portal_url: "{{ portal_url }}"
            portal_key: "{{ portal_key }}"
   ```

### Writing Integration Tests

- Each module must also be added to the all group in meta/runtime.yml. This will allow to reuse common variables like portal_key in the test. 

- All modules MUST have integration tests for new features. Bug fixes for modules that currently have integration tests SHOULD have tests added.

- Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature. The tests are supposed to be added for all writable attributes of the object. The complete list of attributes can be found in the [API Documentation](https://csp.infoblox.com/apidoc).

### Expected test criteria

- Resource creation under check mode
- Resource creation
- Resource creation again (idempotency)
- Resource deletion under check mode
- Resource deletion
- Resource deletion again (idempotency)
- Resource updation with all additional attributes under the object

> For more detailed example, refer to [/tests/integration/targets/ipam_ip_space/tasks/main.yml](../tests/integration/targets/ipam_ip_space/tasks/main.yml)

Reviewer: ```suggestion
For more detailed example, refer to any of test examples(../tests/integration/targets/ipam_ip_space/tasks/main.yml)
```
We can have the link on the examples keyword here.

### Coding Style Guidelines

-  **Properly Naming Tasks** : Each task within your tests should have a specific, descriptive name that clearly states what the test does.
      ```
    - name: "Create an IP space"
      infoblox.universal_ddi.ipam_ip_space:
          name: "{{ name }}"
          state: "present"
          .....
   ``` 

- Use well-defined and easily readable variable names.

     ```name: "test-ip-space"```

- **Clean Up After Tests**: Always clean up test artifacts to ensure tests do not interfere with each other.

    ```yaml
  always:
    # Cleanup if the test fails
    - name: "Delete IP Space"
      infoblox.universal_ddi.ipam_ip_space:
        name: "{{ name }}"
        state: "absent"
      ignore_errors: true
    ```

### Linting

> To facilitate consistent coding style across the project, please run the following command before running the tests

Reviewer: ```suggestion
To facilitate consistent formatting across the project, please run the following command before running the tests
```

# Contributing to infoblox.universal_ddi collection

Welcome to the Universal DDI Ansible Repository ! We are excited that you are interested in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure and coding standards.
## Workflow Summary

1. [Setup](#setup)
2. [Implementing the Code Changes](#implementing-the-code-changes)
3. [Testing the Code Changes](#testing-the-code-changes)
4. [Creating a PR](#creating-a-pr)

## Setup

The `ansible-test` command expects that the repository is in a directory that matches it's collection. Here's the required directory structure you need to set up in your development environment:

```
<target_directory>/ansible_collections/infoblox/universal_ddi/
├── changelogs/
├── docs/
├── meta/
├── plugins/
│   ├── doc_fragments/
│   ├── lookup/
│   ├── module_utils/
│   ├── modules/
│   └── plugin_utils/
├── tests
│   ├── integration/
│   │   └── targets/
│   └── unit/
└── venv/
```

> 

Reviewer: We can mention here that the user should create a fork and the changes should be made to their fork.

Clone the repository from GitHub ensuring the hierarchy.
```shell
mkdir -p $TARGET_DIR/ansible_collections/infoblox

> git clone <url> $TARGET_DIR/ansible_collections/infoblox/universal_ddi

Reviewer: Once the fork line is added , the `<url>` can replaced with the actual URL and we can add a placeholder for user name.

**Return Section** It defines what data is supposed to be returned and how it is structured.

## Testing the Code Changes

### Writing Integration Tests

- Each module must also be added to the all group in meta/runtime.yml. This will allow to reuse common variables like portal_key in the test. 

- All modules MUST have integration tests for new features. Bug fixes for modules that currently have integration tests SHOULD have tests added.

- Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature. The tests are supposed to be added for all writable attributes of the object. The complete list of attributes can be found in the [API Documentation](https://csp.infoblox.com/apidoc).

### Coding Style Guidelines

-  **Properly Naming Tasks** : Each task within your tests should have a specific, descriptive name that clearly states what the test does.
      ```
    - name: "Create an IP space"
      infoblox.universal_ddi.ipam_ip_space:
          name: "{{ name }}"
          state: "present"
          .....
   ``` 

- Use well-defined and easily readable variable names.

     ```name: "test-ip-space"```

> - If your playbook has many debug statements consider using the `verbosity` parameter that allows you to control when certain messages appear.

Reviewer: I feel that this guideline is not required . Under the "Running Tests" , this has already been mentioned and using -vv is better because this way contributor might leave debug statements in code.

  ```
  - debug:
      msg: "This message always appears on the console."

  - debug:
      msg: "This message only appears on the console with ansible-playbook -vv+"
      verbosity: 2
    ```

- **Clean Up After Tests**: Always clean up test artifacts to ensure tests do not interfere with each other.

    ```yaml
  always:
    # Cleanup if the test fails
    - name: "Delete IP Space"
      infoblox.universal_ddi.ipam_ip_space:
        name: "{{ name }}"
        state: "absent"
      ignore_errors: true
    ```

> ### Expected test criteria

Reviewer: This section should be moved before coding style guidelines.

- Resource creation under check mode
- Resource creation
- Resource creation again (idempotency)

> - Resource updation with all additional attributes under the object

Reviewer: Resource updation should come after the line "Resource deletion again (idempotency)"

- Resource deletion under check mode
- Resource deletion
- Resource deletion again (idempotency)

For more detailed example, refer to [/tests/integration/targets/ipam_ip_space/tasks/main.yml](../tests/integration/targets/ipam_ip_space/tasks/main.yml)

> ### Configuration

Reviewer: Configuration Section should come before Writing Integration Tests , under Testing Code Changes

# Contributing to infoblox.universal_ddi collection

Welcome to the Universal DDI Ansible Repository ! We're thrilled about your interest in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure and coding standards.

> ## Workflow Summary

Reviewer: This loogs good but , Better Make this as index only and all the headings , Workflow Summary does not make sense anymore sense we added all the links

- [Testing the Code Changes](#testing-the-code-changes)
    - [Configuration](#configuration)
    - [Writing Integration Tests](#writing-integration-tests)
    - [Expected Test Criteria](#expected-test-criteria)
    - [Coding Style Guidelines](#coding-style-guidelines)
    - [Linting](#linting)
    - [Running Tests](#running-tests)
- [Creating a PR](#creating-a-pr)

## Setup

### Directory Structure

The `ansible-test` command expects that the repository is in a directory that matches it's collection. Here's the required directory structure you need to set up in your development environment:

```
<target_directory>/ansible_collections/infoblox/universal_ddi/
├── changelogs/
├── docs/
├── meta/
├── plugins/
│   ├── doc_fragments/
│   ├── lookup/
│   ├── module_utils/
│   ├── modules/
│   └── plugin_utils/
├── tests
│   ├── integration/
│   │   └── targets/
│   └── unit/
└── venv/
```
Before making any changes, it's essential to fork the repository. This allows you to work on your own copy of the project and makes it easy to contribute your changes back to the main repository through a pull request.

Clone the repository from GitHub ensuring the hierarchy.
```shell

> mkdir -p $TARGET_DIR/ansible_collections/infoblox

Reviewer: ```suggestion
mkdir -p $TARGET_DIR/ansible_collections/infoblox/universal_ddi
```

```

### Installing the Requirements

The python requirements must be installed separately. It is recommended to use a Virtual Environment to isolate the dependencies and maintain a clean setup. 

Setting up the Python virtual environment

```shell
cd $TARGET_DIR/ansible_collections/infoblox/universal_ddi
python3 -m venv .venv
source .venv/bin/activate
```

Installing the requirements

```
pip3 install ansible
pip3 install -r requirements.txt
pip3 install -r test-requirements.txt
```

You can run ```ansible-test sanity``` to test if your collection is ready for development

## Implementing the Code Changes

### Reference Documentation

Before you start contributing, it is essential to familiarize yourself with the [Infoblox API Documentation](https://csp.infoblox.com/apidoc), which provides detailed information on attributes and functionality relevant to the modules you may wish to contribute.

### Developing New Modules
 
Create two files for each module:
- `module_name.py`: Module for performing CRUD operations.
- `module_name_info.py`: Module to List / Read the object.

Place these files under `/plugins/modules`.

#### Documentation, Examples and Return section inside modules

> **Documentation Section** Each module file should start with a comprehensive documentation section. [Module format and documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#module-format-and-documentation) guide gives  details about how to document.

Reviewer: ```suggestion
**Documentation Section** : Each module file should start with a comprehensive documentation section. [Module format and documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#module-format-and-documentation) guide gives  details about how to document.
```

Applies for other two points too.
